signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,571 @@
+    }
+
+    /**
+     * Builder of WebSocket Transport Instance
+     */
+    public static class TransportBuilder {
+
+        private String token;
+        private String protocol = "wss";
+        private String host = SignalFlowTransport.DEFAULT_HOST;
+        private int port = 443;
+        private int timeout = 1;
+        private int version = 2;
+
+        public TransportBuilder(String token) {
+            this.token = token;
+        }
+
+        public TransportBuilder setProtocol(String protocol) {
+            this.protocol = protocol;
+            return this;
+        }
+
+        public TransportBuilder setHost(String host) {
+            this.host = host;
+            return this;
+        }
+
+        public TransportBuilder setPort(int port) {
+            this.port = port;
+            return this;
+        }
+
+        public TransportBuilder setTimeout(int timeout) {
+            this.timeout = timeout;
+            return this;
+        }
+
+        public TransportBuilder setAPIVersion(int version) {
+            this.version = version;
+            return this;
+        }
+
+        public WebSocketTransport build() {
+            SignalFxEndpoint endpoint = new SignalFxEndpoint(this.protocol, this.host, this.port);
+            WebSocketTransport transport = new WebSocketTransport(this.token, endpoint,
+                    this.version, this.timeout);
+            return transport;
+        }
+    }
+
+    /**
+     * Special type of StreamMessage for conveying websocket/connection errors to channels
+     */
+    protected static class SignalFlowExceptionStreamMessage extends StreamMessage {
+
+        protected SignalFlowException exception;
+
+        public SignalFlowExceptionStreamMessage(final SignalFlowException exception) {
+            super("error", null, exception.getMessage());
+            this.exception = exception;
+        }
+
+        public SignalFlowException getException() {
+            return this.exception;
+        }
+    }
+
+    /**
+     * Value Object that handles binary data message conversion in construction
+     */
+    protected static class TransportDataMessage {
+
+        protected byte version;
+        protected Kind kind;
+        protected String channelName;
+        protected long logicalTimestampMs;
+        protected List<Map<String, Object>> data = new ArrayList<Map<String, Object>>();
+
+        public TransportDataMessage(byte[] data) {
+            try {
+                ByteBuffer buffer = ByteBuffer.wrap(data);
+
+                this.version = buffer.get(0);
+                this.kind = Kind.fromBinaryType(buffer.get(1));
+                this.channelName = new String(data, 4, 16, "UTF-8");
+
+                if (this.kind == Kind.DATA) {
+                    this.logicalTimestampMs = buffer.getLong(20);
+
+                    byte[] payload = Arrays.copyOfRange(data, 32, data.length);
+                    buffer = ByteBuffer.wrap(payload);
```

Comment:
Why are you wrapping again? You should be able to use the original buffer, just moving your position around.